[docs] Distributed Snapshots documentation (#12551)
yugabyte/yugabyte-db: docs/content/preview/manage/backup-restore/snapshot-ysql.md

</ul>

The most efficient way to backup the data stored in YugabyteDB is to create a distributed snapshot. A snapshot is a consistent cut of a data taken across all the nodes in the cluster. For YSQL, snapshots are created on per-database level. Backing up individual tables is currently not supported.

When YugabyteDB creates a snapshot, it doesn't physically copy the data, but instead creates hard links to all the relavant files. These links reside on the same storage volumes where the data itself is stored, which makes both backup and restore operations nearly instantanious.

{{< note title="Note on space consumption" >}}

> There are no technical limitations on how many snapshots you can create. However, it's important to keep in mind that by increasing the number of snapshots stored, you also increase the amount of space required for the database. The actual overhead depends on the workload, so we recommend to estimate it by running tests based on your applications.

Reviewer: Any recommendations on what tests to run, and how, or is that going to be super-specific to each use case?
```suggestion
There are no technical limitations on how many snapshots you can create. However, increasing the number of snapshots stored also increases the amount of space required for the database. The actual overhead depends on the workload, but you can estimate it by running tests based on your applications.
```

{{< /note >}}

## Creating a Snapshot

The distributed snapshots feature allows you to backup a database, and then restore it in case of a software or operational error, with minimal RTO and overhead.

> To backup a database, create a snapshot using the [`create_database_snapshot`](../../../admin/yb-admin/#create-database-snapshot) command:

Reviewer: ```suggestion
To back up a database, create a snapshot using the [`create_database_snapshot`](../../../admin/yb-admin/#create-database-snapshot) command:
```

```sh
yb-admin create_database_snapshot my_database
```

> When the command is executed, it generates a unique ID for the snapshot, and prints it out:

Reviewer: Avoid passive voice
```suggestion
When you run the command, it returns a unique ID for the snapshot:
```

```output
Started snapshot creation: 0d4b4935-2c95-4523-95ab-9ead1e95e794
```

You can then use this ID to check the status of the snapshot, [delete it](#deleting-a-snapshot), or use it to [restore the database](#restoring-a-snapshot).

> The `create_database_snapshot` command exits immidiately, so its completion does not necessarily means that the snapshot is successfully created. Before using the snapshot, you should verify its status by running the [`list_snapshots`](../../../admin/yb-admin/#list-snapshots) command:

Reviewer: ```suggestion
The `create_database_snapshot` command exits immediately, but the snapshot may take some time to complete. Before using the snapshot, verify its status with the [`list_snapshots`](../../../admin/yb-admin/#list-snapshots) command:
```

```sh
yb-admin list_snapshots
```

This command will print out all the snapshots that exist in the cluster with their statuses. Locate the ID of the new snapshot and make sure its status is `COMPLETE`:

```output
Snapshot UUID                         State
0d4b4935-2c95-4523-95ab-9ead1e95e794  COMPLETE
```

> ## Deleting a Snapshot

Reviewer: ```suggestion
## Delete a snapshot
```

> Snapshots created using the `create_database_snapshot` never expire and are retained as long as the cluster exists. If a snapshot is no longer needed, you can delete it by running the [`delete_snapshot`](../../../admin/yb-admin/#delete-snapshot) command and providing the ID of the snapshot:

Reviewer: ```suggestion
Snapshots created using the `create_database_snapshot` never expire and are retained as long as the cluster exists. If you no longer need a snapshot, you can delete it with the [`delete_snapshot`](../../../admin/yb-admin/#delete-snapshot) command:
```

```sh
yb-admin delete_snapshot 0d4b4935-2c95-4523-95ab-9ead1e95e794
```

> ## Restoring a Snapshot

Reviewer: ```suggestion
## Restore a snapshot
```

> To restore a database to a previously created snapshot, run the [`restore_snapshot`](../../../admin/yb-admin/#restore-snapshot) command and provide the ID of the snapshot:

Reviewer: Again, terminology: the tool calls this a UUID, not an ID.

```sh
yb-admin restore_snapshot 0d4b4935-2c95-4523-95ab-9ead1e95e794
```

> The above command will rollback the database to the state which it had when the snapshot was created. The restore happens in-place, i.e. it changes the state of the existing database within the same cluster.

Reviewer: No Latin terms.

**CAUTION** "roll back" has a very specific meaning. Are you sure this term is correct here? I've been smacked for this one before. :)
```suggestion
This command rolls back the database to the state which it had when the snapshot was created. The restore happens in-place: in other words, it changes the state of the existing database within the same cluster.
```

> ## Moving a Snapshot to an External Storage

Reviewer: ```suggestion
## Move a snapshot to external storage
```

> Storing snapshots in-cluster is extermely efficient, but also comes with downsides. First of all, it increases the cost of the cluster - increasing number of snapshot can inflate the space consumption on the storage volumes. Second of all, in-cluster snapshots do not protect you from disaster scenarios like filesystem corruption or hardware failures.

Reviewer: ```suggestion
Storing snapshots in-cluster is extremely efficient, but also comes with downsides. It can increase the cost of the cluster by inflating the space consumption on the storage volumes. Also, in-cluster snapshots don't protect you from a disaster scenario like filesystem corruption or a hardware failure.
```

> To mitigate the above, you might want to store backups outside of the cluster, in a cheaper storage that is also geografically separated from the cluster. This way, you can reduce the cost, and also restore your databases into a different cluster, potentially in a different location.

Reviewer: ```suggestion
To mitigate these issues, consider storing backups outside of the cluster, in cheaper storage that is also geographically separated from the cluster. This approach helps you to reduce the cost, and also to restore your databases into a different cluster, potentially in a different location.
```

> To move a snapshot to an external storage, you need to gather all the relevant files from all the nodes, and copy then along with additional metadata that will be required when you decide to restore the snapshot on a different cluster. Below is the detailed step-by-step explanation of the process.

Reviewer: ```suggestion
To move a snapshot to external storage, gather all the relevant files from all the nodes, and copy them along with the additional metadata required for restores on a different cluster:
```

1. Get the current YSQL schema catalog version by running the [`ysql_catalog_version`](../../../admin/yb-admin/#ysql-catalog-version) command:

    ```sh
    yb-admin ysql_catalog_version
    ```

    ```output

>     Version:1

Reviewer: Nope, the space is there in the output
```suggestion
    Version: 1
```

    ```

> 2. [Create an in-cluster snapshot](#creating-a-snapshot).

Reviewer: Use `1.` for every list item. It makes maintenance easier, and makes us not look like idiots when we repeat number 4 in a list, or jump from 3 to 5. :)

Reviewer: Also, where's the command for this step?

Author: It links to the section that describes how to create a snapshot. It's not a single command.

> 3. Create a backup of the YSQL metadata using the [`ysql_dump`](../../../admin/ysql-dump) command:

Reviewer: ```suggestion
3. Back up the YSQL metadata using the [`ysql_dump`](../../../admin/ysql-dump) command:
```

Reviewer: Don't we have a new approach to backups that doesn't involve dumps now? I thought we did...

    ```sh
    ysql_dump --include-yb-metadata --serializable-deferrable --create --schema-only --dbname my_database --file my_database_schema.sql
    ```

> 4. Repeat step #1 and verify that the catalog version is the same as it was prior to creating the snapshot. If it is not, you're not guaranteed to get a consistent restorable snapshot, and should restart the process.

Reviewer: ```suggestion
4. Verify that the catalog version is the same as it was prior to creating the snapshot. If it isn't, you're not guaranteed to get a consistent restorable snapshot, and should restart the process.

    ```sh
    yb-admin ysql_catalog_version
    ```
```

    <yb_data_dir>/node-<node_number>/disk-<disk_number>/yb-data/tserver/data/rocksdb/table-<table_id>/[tablet-<tablet_id>.snapshots]/<snapshot_id>
    ```

>     - `<yb_data_dir>` is the directory where YugabyteDB data is stored. Default is `~/yugabyte-data`.

Reviewer: ```suggestion
    - `<yb_data_dir>` is the directory where YugabyteDB data is stored. The default value is `~/yugabyte-data`.
```

    {{< /note >}}

8. If you don't want to keep the in-cluster snapshot, it's now safe to [delete it](#deleting-a-snapshot).

> ## Restoring a Snapshot from an External Storage

Reviewer: ```suggestion
## Restore a snapshot from external storage
```

> To restore a snapshot that had been [moved to an external storage](#moving-a-snapshot-to-an-external-storage), go through the steps below.

Reviewer: After you've cleaned up all the headings, make sure that all the anchor links still work...
```suggestion
To restore a snapshot that you've [moved to external storage](#move-a-snapshot-to-external-storage), do the following:
```

1. Fetch the YSQL metadata file from the external storage and apply it using the [`ysqlsh`](../../../admin/ycqlsh/) CLI tool:

    ```sh
    ysqlsh -h 127.0.0.1 --echo-all --file=my_database_schema.sql
    ```

2. Fetch the snapshot metadata file from the external storage and apply it by running the [`import_snapshot`](../../../admin/yb-admin/#import-snapshot) command:

    ```sh
    yb-admin import_snapshot my_database.snapshot my_database
    ```

>     The output will contain the mapping between the old tablet IDs and the new tablet IDs:

Reviewer: ```suggestion
    The output contains the mapping between the old tablet IDs and the new tablet IDs:
```

    Snapshot         0d4b4935-2c95-4523-95ab-9ead1e95e794   6beb9c0e-52ea-4f61-89bd-c160ec02c729
    ```

> 3. Copy the tablet snapshots.

Reviewer: `1.` numbering for all list items...

>     Use the tablet mappings to copy the tablet snapshot files from the external location to appropriate location.

Reviewer: ```suggestion
    Use the tablet mappings to copy the tablet snapshot files from the external storage to the appropriate location.
```

    {{< /note >}}

4. [Restore the snapshot](#restoring-a-snapshot).

> -----

Reviewer: Why the horizontal rule here? They are almost never necessary.
```suggestion
```

> {{< note title="YugabyteDB Anywhere Automated Backups" >}}

Reviewer: ```suggestion
{{< note title="Automated backups for YugabyteDB Anywhere" >}}
```

> YugabyteDB Anywhere provides the API and UI for [Backup and Restore](../../../yugabyte-platform/back-up-restore-universes/), which automates most of the steps described above. Consider using it for better usability, especially if you have many databases and snapshots to manage.

Reviewer: ```suggestion
YugabyteDB Anywhere provides an API and UI for [backup and restore](../../../yugabyte-platform/back-up-restore-universes/), which automate most of the steps described here. You should use one or both, especially if you have many databases and snapshots to manage.
```
